### ISSUE TITLE
[no-throw-literal] assumes intersection types are not errors

### ISSUE THREAD

**Issue by martinheidegger** [typescript-eslint/typescript-eslint#1841]
> Make sure you read through our FAQ before posting.
> https://github.com/typescript-eslint/typescript-eslint/blob/issue-template-update/docs/getting-started/linting/FAQ.md

_Note:_ ↑ Link is broken, I read [this](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md).

**Repro**

_.eslintrc_
```javascript
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: { project: './tsconfig.json' },
  plugins: [ '@typescript-eslint' ],
}
```

_test.ts_
```TS
/*eslint @typescript-eslint/no-throw-literal: "error"*/
function foo () {
  throw Object.assign(new Error('message'), { foo: 'bar' })
}
```

**Expected Result**

No error, as the union type of `Error | { foo: 'bar' }` does implement "Error"

**Actual Result**

```
  3:9  error  Expected an error object to be thrown  @typescript-eslint/no-throw-literal
```

**Additional Info**

This is somewhat related, but in typescript `Foo & Bar` specifies the api but doesn't specify how `instanceof` is supposed to behave. `Foo & Error` could be an instance of `Foo` or `Error` or something completely different. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.26.0` |
| `@typescript-eslint/parser`        | `2.26.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.11.11` |
| `npm`                              | `6.14.1` |

**Comment by bradzacher** [comment]
Intersection types just haven't been implemented in the rule.
They need special handling, unfortunately.
Happy to accept a PR.

**Comment by a-tarasyuk** [comment]
I suppose need to handle only `&` not, `|`. Am I right?

**Comment by bradzacher** [comment]
Intersection `&` types are a must, yeah.
Union `|` types should be valid if and only if every type in the union is a valid error type.

`Error & { bar: string }` is valid
`Error | SyntaxError` is valid

`Error | { bar: string }` is invalid.
